hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieBaseRelation.scala

```diff
@@ -286,8 +319,16 @@ abstract class HoodieBaseRelation(val sqlContext: SQLContext,
   }
 
   protected final def appendMandatoryColumns(requestedColumns: Array[String]): Array[String] = {
-    val missing = mandatoryColumns.filter(col => !requestedColumns.contains(col))
-    requestedColumns ++ missing
+    if (dropPartitionColumnsWhenWrite) {
+      if (requestedColumns.isEmpty) {
+        mandatoryColumns.toArray
+      } else {
+        requestedColumns
```

Comment:
why not requestedColumns ++ missing ?

Reply:
Using `requestedColumns ++ missing` will involve other columns which don't be needed, and this does not work well alone with the partition columns provided. The phenomenon is that the value of an extra useless fields replaces the values of the partition fields.
In the case, I test and notice that will return list of `InternalRow`, inside of which the extra, useless columns will be put between  the required data columns and partition columns. I guess the first n values will be returned, instead of return values by schema.
Anyway, the logical above works well in all cases, and don't take more overhead.

Reply:
@YannByron let's make sure we're carefully vetting changes we're making on that level, since i'd imagine we didn't do exhaustive testing with and without this new config.

A few notes here: 

1. Behavior of selecting mandatory columns should not be predicated on this config
2. This change will break MOR tables semantic (when new config is on) as it misses required columns for MOR to do merging properly

Reply:
1. actually, i can't understand why need to query additional mandatory columns in all the cases. If no column is required, we can fetch recordkey, and for mor merge, recordkey and precombineField. IMO, this is enough.

2. if so, i can fix it. or just forbid enable this setting for mor table. after all, i don't think this setting is reasonable.

Reply:
@YannByron set of required fields varies b/w different tables relying on making sure that we at least fetch fields to appropriate handle table's semantic. In MOR case such mandatory fields are exactly the record-key and pre-combine-key.